### PR TITLE
[llama2] Use invoke_dynamic in tensor filter llama2

### PR DESCRIPTION
**Pull request by niley7464**
This patch adds invoke_dynamic funciton to tensor filter llama2.
LLM model's output is flexible, so llama2 filter should use invoke_dynamic.

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

**Comment by niley7464** [comment]
https://github.com/nnsuite/llama2.c/blob/0804e8a19f747225f3f92aeda02b480974763439/transformer.c#L13
```c
#define PRAGMA_OMP_PARALLEL_FOR_PRIVATE(x) DO_PRAGMA(omp parallel for private(x))
```

This part causes 
```
Freeing pipeline ...
Caught SIGSEGV
Caught SIGSEGV
Caught SIGSEGV
Caught SIGSEGV
Caught SIGSEGV
Caught SIGSEGV
```

Tested with no openmp-support
```
# features
option('openmp-support', type: 'feature', value: 'disabled')
```

File: ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc [hunk 1]
```diff
@@ -165,17 +182,24 @@ TensorFilterLlama2c::invoke (const GstTensorMemory *input, GstTensorMemory *outp
 
   result = get_tokens (transformer, tokenizer, sampler, prompt, steps);
 
-  /** @todo Make out_info flexible */
-  output[0].size = tokenizer->max_token_length * steps;
+  if (!result) {
```
Comment (jaeyun-jung) [inline ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc:185]: Q: is it possible if returned token is empty ?
```
if (!result || (len = strlen (result)) == 0)
```

Comment (niley7464) [reply ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc:185]: `result`'s memory is allocated in llama2c, so it must be freed even if length is 0.
I added a logic to check if string is empty!
Also, I found `prompt`'s memory is not relased by your review.
Thank you :)

**Review by jaeyun-jung — APPROVED** [review]

File: ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc [hunk 2]
```diff
@@ -15,13 +15,14 @@
  * @details Pipeline example
  *          gst-launch-1.0 filesrc location=input.txt ! application/octet-stream !
  *          tensor_converter ! other/tensors,format=flexible !
- *          tensor_filter framework=llama2c model=model.bin,tokenizer.bin !
- *          filesink location=output.txt
+ *          tensor_filter framework=llama2c model=model.bin,tokenizer.bin invoke-dynamic=TRUE !
+ *          application/octet-stream ! filesink location=output.txt
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc:19]: Is if ok to "decode" ```other/tensors``` to ```application/octet-stream``` with caps-filter?

File: ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc [hunk 3]
```diff
@@ -147,7 +150,21 @@ TensorFilterLlama2c::configure_instance (const GstTensorFilterProperties *prop)
...
+}
+
+/**
+ * @brief Invoke llama2c using input tensors
+ */
+void
+TensorFilterLlama2c::invoke_dynamic (GstTensorFilterProperties *prop,
+    const GstTensorMemory *input, GstTensorMemory *output)
+{
+  char *result = nullptr, *prompt;
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc:167]: Be careful. ```prompt``` is also a class property.
Using same name for a local var and a class property is not good.

File: ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc [hunk 4]
```diff
@@ -83,7 +86,7 @@ class TensorFilterLlama2c : public tensor_filter_subplugin
   static Sampler *sampler;
   static Config *config;
   char *prompt; // prompt string
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc:88]: This ```prompt``` property is not used...

File: ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc [hunk 5]
```diff
@@ -165,17 +182,32 @@ TensorFilterLlama2c::invoke (const GstTensorMemory *input, GstTensorMemory *outp
 
   result = get_tokens (transformer, tokenizer, sampler, prompt, steps);
 
-  /** @todo Make out_info flexible */
-  output[0].size = tokenizer->max_token_length * steps;
-  output[0].data = allocate (output[0].size);
+  if (!result) {
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc:185]: LINE 168-181 == strndup

Comment (myungjoo) [reply ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc:185]: LINE 174 implies that the input tensor[0] is a char string ending with '0'. Is this a safe assumption?

File: ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc (hunk 2 above)
Comment (myungjoo) [reply ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc:19]: And.. what happens if in/out tensors are format=static?

Comment (niley7464) [reply ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc:20]: Oh, I missed tensor_decoder line 😅 Updated it!
This filter is for text generation, so if the tensor_filter's invoke-dynamic option is FALSE (in/out static) it thorws runtime error. 
Same with #4606

File: ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc (hunk 4 above)
Comment (niley7464) [reply ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc]: Removed unused `prompt`
Thank you :)

File: ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc (hunk 5 above)
Comment (niley7464) [reply ext/nnstreamer/tensor_filter/tensor_filter_llama2.cc:184]: Used `strndup` for input and output data :)
'\0' is arbitrarily added at LINE 171 to use `strlen`.

**Review by anyj0527 — APPROVED** [review]
LGTM👍

**Review by myungjoo — APPROVED** [review]

**Pull request merged**